net/src/packet.rs

```diff
@@ -248,6 +260,199 @@ impl DeParse for Packet {
                 cursor.write(transport)?;
             }
         }
+
+        match self.encap {
+            None => {
+                return Ok(NonZero::new(cursor.inner.len() - cursor.remaining)
+                    .unwrap_or_else(|| unreachable!()))
+            }
+            Some(Encap::Vxlan(ref vxlan)) => {
+                cursor.write(vxlan)?;
+            }
+        }
         Ok(NonZero::new(cursor.inner.len() - cursor.remaining).unwrap_or_else(|| unreachable!()))
     }
 }
+
+#[derive(Debug, thiserror::Error)]
+#[error("Packet already has as many VLAN headers as parser can support (max is {MAX_VLANS})")]
+pub struct TooManyVlans;
+
+impl Packet {
+    /// Create a new packet with the supplied `Eth` header.
+    fn new(eth: Eth) -> Packet {
```

Comment:
Nice, I've been missing this for my tests with NAT :+1:

Reply:
That makes sense.

Hopefull the follow on PRs will make this passably ergonomic.

I have plans to use tools like bolero and smoltcp to make this much more test friendly in the future but for now we have packet generation as something of a pain point